src/util/AllocatorWithLimit.h

```diff
@@ -173,32 +177,37 @@ class AllocatorWithLimit {
     const auto bytesNeeded = n * sizeof(T);
     const bool wasEnoughLeft =
         memoryLeft_.ptr()->wlock()->decrease_if_enough_left_or_return_false(
-            bytesNeeded);
+            MemorySize::bytes(bytesNeeded));
     if (!wasEnoughLeft) {
       clearOnAllocation_(n);
```

Comment:
We should also have a look at the semantics of `clearOnAllocation_`, maybe it should also take a `MemorySize` to make it more clear (have a look at how it's used for the cache)

Reply:
Alright, as far as I **can tell** (There is no documentation, except the name and this one use.) this function is supposed to free memory and take the same argument as [`std::allocator<T>::allocate`](https://en.cppreference.com/w/cpp/memory/allocator/allocate).

Which would make using `MemorySize` as horrendous choice, because `std::allocator<T>::allocate` takes the number of **objects**, not bytes.

Of course, we could just adjust the input, but I'm not sure, because the original version could be clearer in it's use, in combination with `allocate`.

P.S.: If my conclusion about the current correct usage is correct, then the custom function for `clearOnAllocation_` in `Server.cpp` (Constructor of the `Server` class.) is just plain wrong and needs to be fixed pronto.

Reply:
I just checked, and it should use `MemorySize`. (`allocate()` should in principle also take `MemorySize`, but it can't, because it has to interact with the standard library).

And if you mean the bug that we currently divide by `sizeof(Id)` in the Server.cpp, yes, that looks wrong, please fix it.
Also a good argument to make the ClearOn...Function take MemorySize: then this call also becomes much simpler .

Reply:
> I just checked, and it should use `MemorySize`. (`allocate()` should in principle also take `MemorySize`, but it can't, because it has to interact with the standard library).

Are you sure? The type of object, and it's size, is always the same and known by the allocator, so wouldn't the need, to always repeat it in the call, be very annoying?

> And if you mean the bug that we currently divide by `sizeof(Id)` in the Server.cpp, yes, that looks wrong, please fix it.

Alright, will do.

> Also a good argument to make the ClearOn...Function take MemorySize: then this call also becomes much simpler .

I do not follow. The call would become more complex, because it would have to use `byteNeeded` instead of `n`. (Which would be more complex, if we didn't already calculate it.) And it would also no longer work the same as the standard, which will lead to confusion.

Reply:
Fixed the big in `Server.cpp`. See [371bef0](https://github.com/ad-freiburg/qlever/pull/1112/commits/371bef0643ddad0634f5430dbc97f7f50850a6ca).

Reply:
Changed `clearOnAllocation_`. See [ae4a8ab](https://github.com/ad-freiburg/qlever/pull/1112/commits/ae4a8abe4503d17cdb9bbfb75b5c87331f9f42f5).